x-pack/legacy/plugins/monitoring/server/lib/logstash/get_paginated_pipelines.js

```diff
@@ -45,61 +46,155 @@ export async function getPaginatedPipelines(
+  }
+
+  // Filtering
+  const filteredPipelines = filter(pipelines, queryText, ['id']); // We only support filtering by id right now
+
+  // Sorting
+  const sortedPipelines = sortPipelines(filteredPipelines, sort);
+
+  // Pagination
+  const pageOfPipelines = paginate(pagination, sortedPipelines);
+
+  const response = {
+    pipelines: await getPipelines(
+      req,
+      lsIndexPattern,
+      pageOfPipelines,
+      throughputMetric,
+      nodesCountMetric
+    ),
+    totalPipelineCount: filteredPipelines.length,
+  };
+
+  return processPipelinesAPIResponse(response, throughputMetric, nodesCountMetric);
+}
+
+function processPipelinesAPIResponse(response, throughputMetricKey, nodesCountMetricKey) {
+  // Clone to avoid mutating original response
+  const processedResponse = cloneDeep(response);
+
+  // Normalize metric names for shared component code
+  // Calculate latest throughput and node count for each pipeline
+  processedResponse.pipelines.forEach(pipeline => {
+    pipeline.metrics = {
+      throughput: pipeline.metrics[throughputMetricKey],
+      nodesCount: pipeline.metrics[nodesCountMetricKey],
+    };
+
+    pipeline.latestThroughput = last(pipeline.metrics.throughput.data)[1];
+    pipeline.latestNodesCount = last(pipeline.metrics.nodesCount.data)[1];
+  });
+  return processedResponse;
+}
+
+async function getPaginatedThroughputData(pipelines, req, lsIndexPattern, throughputMetric) {
   const metricSeriesData = Object.values(
     await Promise.all(
       pipelines.map(pipeline => {
         return new Promise(async resolve => {
           const data = await getMetrics(
             req,
             lsIndexPattern,
-            metricSet,
+            [throughputMetric],
             [],
             {
               pipeline,
             },
             2
           );
-
-          resolve({
-            id: pipeline.id,
-            metrics: Object.keys(data).reduce((accum, metricName) => {
-              accum[metricName] = data[metricName][0];
-              return accum;
-            }, {}),
-          });
+          resolve(reduceData(pipeline, data));
         });
       })
     )
   );
+
   for (const pipelineAggregationData of metricSeriesData) {
     for (const pipeline of pipelines) {
       if (pipelineAggregationData.id === pipeline.id) {
-        for (const metric of metricSet) {
-          const dataSeries = get(pipelineAggregationData, `metrics.${metric}.data`, [[]]);
-          pipeline[metric] = dataSeries[dataSeries.length - 1][1];
-        }
+        const dataSeries = get(pipelineAggregationData, `metrics.${throughputMetric}.data`, [[]]);
+        pipeline[throughputMetric] = dataSeries.pop()[1];
       }
     }
   }
+}
 
-  // Manually apply pagination/sorting/filtering concerns
+async function getPaginatedNodesData(pipelines, req, lsIndexPattern, nodesCountMetric) {
+  const metricSeriesData = await getMetrics(req, lsIndexPattern, [nodesCountMetric], [], {}, 2);
```

Comment:
Can we pass in `pipelines` here so we can reduce the number of buckets created in [these aggs](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/monitoring/server/lib/metrics/logstash/classes.js#L312)?